### DISCUSSION TITLE
custom color requires important everywhere

### DISCUSSION THREAD

**Discussion by alicerocheman** [tailwindlabs/tailwindcss#5279] [Help] [answered]
Hello,

I have a Next app using Tailwind JIT.

I have a custom color called "primary" set in the tailwind config, which has been working as expected.

Recently, everywhere I used `text-primary` stopped working wherever it was overloading a `text-gray-`, forcing me to replace them all with `!text-primary`.

It used to work properly and I have no idea why all of a sudden my elements were gray. It's like my "primary" custom color has somehow lost priority.

I'm wondering if it's just me or if maybe someone else had the same issue, or if someone knows what we might have done or how to fix it. We haven't touched the color config recently. 

Any help would be appreciated.

## Expected (and the way it used to work until recently)
button text is displayed in the primary color

## Current behavior
if no `!` is added to `text-primary` the button text is gray.

## Usage
```tsx
const ComponentA:FC<WithClassName> = ({className}) => {
  const computedClassName = useMemo(() => `text-gray-500 ${className ?? ''}`, [className]);
  return <button className={computedClassName}>Button</button>
}

const ComponentB:FC = () => {
  return <ComponentA className="text-primary" />
}
```tsx

## Config
```js
/** @typedef { import('tailwindcss/defaultConfig')} DefaultConfig */
/** @typedef { import('tailwindcss/defaultTheme')} DefaultTheme */
/** @typedef { DefaultConfig & {theme:{extend:DefaultTheme}}} TailwindConfig */
const defaultTheme = require('tailwindcss/defaultTheme');
const safePaddingPlugin = require('./styles/safe-padding-plugin');

const sizes = {
  15: '3.75rem',
  17: '4.25rem',
  18: '4.5rem',
  30: '7.5rem',
  68: '17rem',
  75: '18.75rem',
  79: '19.75rem',
  86: '21.5rem',
  90: '22.5rem',
  100: '25rem',
  112: '28rem',
  128: '32rem',
  144: '36rem',
  160: '40rem',
  192: '48rem',
};

/** @type {TailwindConfig} */
module.exports = {
  mode: 'jit',
  presets: [require('tailwindcss/defaultConfig')],
  purge: [
    './pages/**/*.{js,ts,jsx,tsx}',
    './components/**/*.{js,ts,jsx,tsx}',
    './lib/**/*.{js,ts,jsx,tsx}',
    './hooks/**/*.{js,ts,jsx,tsx}',
  ],
  darkMode: false, // or 'media' or 'class'
  theme: {
    screens: {
      xxs: '350px',
      xs: '520px',
      sm: '640px', //default
      md: '768px', //default
      lg: '1024px', //default
      lgxl: '1152px',
      xl: '1280px', //default
      '2xl': '1536px', //default
    },
    boxShadow: {
      ...defaultTheme.boxShadow,
      tooltip: '2px -2px 30px -10px rgba(0,0,0,0.25)',
      modal: '2px -2px 50px -12px rgba(0,0,0,0.25)',
    },
    rotate: {
      '-180': '-180deg',
      '-135': '-135deg', // custom
      '-90': '-90deg',
      '-45': '-45deg',
      '-12': '-12deg',
      '-6': '-6deg',
      '-3': '-3deg',
      '-2': '-2deg',
      '-1': '-1deg',
      0: '0',
      1: '1deg',
      2: '2deg',
      3: '3deg',
      6: '6deg',
      12: '12deg',
      45: '45deg',
      90: '90deg',
      135: '135deg', // custom
      180: '180deg',
    },
    extend: {
      colors: {
        primary: '#1AAE94',
        'primary-dark': '#206A5D',
        'review-yellow': '#FFAA00',
        'private-blue': '#184D7B',
        'info-blue': '#4280FC',
        error: '#AD2216',
        gray: {
          50: '#F5F5F5',
          90: '#EFEFEF',
          100: '#EEEEEE',
          120: '#E9E9E9',
          200: '#DDDDDD',
          210: '#DCDCDC',
          230: '#D8D8D8',
          300: '#CCCCCC',
          400: '#999999',
          500: '#666666',
          700: '#333333',
        },
      },
      flexGrow: {
        2: 2,
      },
      fontFamily: {
        sans: ['"Open Sans"', ...defaultTheme.fontFamily.sans],
      },
      fontSize: {
        xxs: ['.625rem', { lineHeight: '0.75rem' }],
        '2.5xl': ['1.6875rem', { lineHeight: '2.125rem' }],
      },
      height: sizes,
      maxHeight: sizes,
      outline: {
        primary: '2px solid #1AAE94',
      },
      borderWidth: {
        5: '5px',
      },
      inset: {
        26: '6.5rem',
      },
      lineHeight: {
        7.5: '1.875rem',
      },
    },
  },
  variants: {
    width: ['first'],
  },
  plugins: [safePaddingPlugin()],
};
```

**Comment by simonswiss** [comment, marked as answer]
Hey!

In the second line of your code example:

```js
const computedClassName = useMemo(() => `text-gray-500 ${className ?? ''}`, [className]);
```

I can see that you are adding the custom class (`text-primary`) _after_ the `text-gray-500` class.

**Important:** the order in which the classes are applied to an element **do not decide** which one takes precedence.

Both those classes have the same "weight" in terms of specificity, and what will determine which one "wins" is the _source order_ in the generated CSS.

The class that is present _last_ in the CSS output is the one that will prevail, and override the other.

Inverting the order of the classes in the HTML element does not change anything - these classes still both point to a selector in the CSS stylesheet.

In other words, if your base component has a text-color utility class applied, you cannot assume that you can override that color by setting another text-color utility class when "consuming" the component.

That's just how CSS works. Of course when you add the `!` flag, you change the specificity of that class, which now definitely wins the battle. 

I am not 100% sure about this, but I _think_ that the JIT engine will generate the color utilities in the order you've used them in your markup – so if you had used the `text-primary` color first, and then later written a `text-gray-500` class, that `text-gray-500` class would be generated later in the CSS file, and therefore have more specificity than the `text-primary` one.

Anyway, you should not attempt to use two utilities for the same property on the same element. Instead, you could have a `textColor` prop, which defaults to `text-gray-500` if nothing is passed to the `<ComponentA />`, but can be whatever text-color class you pass to that component.

This of course requires a bit more work, because you need to _identify_ the overrides you want to allow instead of letting consumers of the component apply any utility class they want.
